### ISSUE TITLE
What are the consequences of this kernel change?

### ISSUE THREAD

**Issue by andrewbird** [stsp/dosemu2#77]
Buried in this pull request for kernel 4.3 are some changes relating to Dosemu 32bit and cpuemu=off mode. Is this effectively killing that? I think the default was to allow it, but there are some scary warnings about it being a security threat to have it enabled, so I expect vendors will disable by default. Are there any alternatives to recompiling the kernel to work around this?

http://lkml.iu.edu/hypermail/linux/kernel/1508.3/04823.html

**Comment by stsp** [comment]
I don't see what you mean.
Where's the security threat mentioned?

**Comment by andrewbird** [comment]
In the middle of the diff at arch/x86/Kconfig
~~~
+ This option allows user programs to put the CPU into V8086
+ mode, which is an 80286-era approximation of 16-bit real mode.
+
+ Some very old versions of X and/or vbetool require this option
+ for user mode setting. Similarly, DOSEMU will use it if
+ available to accelerate real mode DOS programs. However, any
+ recent version of DOSEMU, X, or vbetool should be fully
+ functional even without kernel VM86 support, as they will all
+ fall back to (pretty well performing) software emulation.
+
+ Anything that works on a 64-bit kernel is unlikely to need
+ this option, as 64-bit kernels don't, and can't, support V8086
+ mode. This option is also unrelated to 16-bit protected mode
+ and is not needed to run most 16-bit programs under Wine.
+
~~~
~~~
+ Enabling this option adds considerable attack surface to the
+ kernel and slows down system calls and exception handling.
~~~
~~~
+ Unless you use very old userspace or need the last drop of
+ performance in your real mode DOS games and can't use KVM,
+ say N here.
~~~

**Comment by stsp** [comment]
Nice...
OK, I mailed them some rant (you are in CC) -
this is all I can do.

**Comment by andrewbird** [comment]
Thanks




Sent from Samsung tablet

-------- Original message --------
From Stas Sergeev <notifications@github.com> 
Date: 02/09/2015  10:34  (GMT+00:00) 
To stsp/dosemu2 <dosemu2@noreply.github.com> 
Cc Andrew Bird <ajb@spheresystems.co.uk> 
Subject Re: [dosemu2] What are the consequences of this kernel change? (#77) 
 
Nice...
OK, I mailed them some rant (you are in CC) -
this is all I can do.

**Comment by stsp** [comment]
Thanks for finding this.
This was so unexpected that I didn't even
believe you initially.

**Comment by andrewbird** [comment]
I'm glad you sorted it out, thanks.

**Comment by stsp** [comment]
At the end it will remain disabled.
In fact, advocating it, I am hiding the fact that
I do not even use it myself, so... this battle is lost.

**Comment by stsp** [comment]
I created the vm86old branch where I reduced
the use of vm86 to the bare minimum.
Could you please test that branch on 32bit?
Linus said that he'll probably allow the minimal
vm86() implementation if someone writes it.

**Comment by andrewbird** [comment]
Unfortunately this branch doesn't startup fully. Dosemu.bin ends up 100% on CPU

Starting terminal
~~~
ajb@polly:/clients/common/dosemu2.git$ /opt/dosemu2/bin/dosemu -X
ERROR: vm86 service not available in your kernel
ERROR: using CPU emulation for vm86()
ALSA lib rawmidi_hw.c:233:(snd_rawmidi_hw_open) open /dev/snd/midiC0D0 failed: No such file or directory
~~~

Last bit of boot.log
~~~
initial register values: fs: 0x0000  gs: 0x0033 eflags: 0x0217                  
PKT: Cannot open tap%d: Operation not permitted                                 
ERROR: vm86 service not available in your kernel                                
ERROR: using CPU emulation for vm86()                                           
CONF: reserving 8256Kb at 0x100000 for 'x' (Extended memory (HMA+XMS))          
Registering HWRAM, type=e base=0xb4fa6000 size=0x400000                         
CONF: reserving 4096Kb at 0xB4FA6000 for 'e' (VGAEMU LFB)                       
CONF: reserving 12Kb at 0xC0000 for 'V' (VGAEMU Video BIOS)                     
SERIAL $Id$                                                                     
CONF: detected layout is "uk"                                                   
CONF: detected alternate layout: fr-latin1                                      
CONF: reserving 16Kb at 0xE4000 for 'E' (EMS page frame)                        
CONF: reserving 16Kb at 0xE8000 for 'E' (EMS page frame)                        
CONF: reserving 16Kb at 0xEC000 for 'E' (EMS page frame)                        
CONF: reserving 16Kb at 0xF0000 for 'E' (EMS page frame)                        
CONF: reserving 132Kb at 0xC3000 for 'U' (Upper Memory Block (UMB, XMS 3.0))    
TIME: using 9154 usec for updating ALRM timer                                   
======================= ENTER CPU-EMU ===============                           
~~~

**Comment by andrewbird** [comment]
So, regarding the kernel change, I thought Linus was going to have vm86 mode disabled at run time if mmap addr != 0. I can work with that it's no different to the setup we have to use now. Has his view changed? I didn't see any more traffic on lkml

**Comment by stsp** [comment]
Thanks for checking.
I'll fix it up later.
So, as you can see, the troubles with vm86()
are big unless someone who really uses it, steps
up to do the work. :)

> So, regarding the kernel change, I thought Linus was going to have vm86 mode
> disabled at run time if mmap addr != 0.

Yes, he did. It is now run-time disabled.

> Has his view changed?

No, its just he never said disabling it run-time will
mean enabling it compile-time. :) We assumed that,
but the result is that it is now disabled both ways. :-P

**Comment by stsp** [comment]
Someone stripped the CC, the thread was CCed only to
some x86@kernel.org, and is not googlable yet.

**Comment by stsp** [comment]
Should now work, except for dosdebug.

**Comment by andrewbird** [comment]
Yes this works fine and has no impact on either my floating-point or integer benchmarks.

**Comment by stsp** [comment]
How does the cpu-emu affects your test-cases?
I guess we'll have to start optimizing cpu-emu.
Another possibility is to write an ultra-ligh and
reasonably compatible vm86() - kernel guys
(including Linus) said they'll accept that and
turn back to Y if it looks good. I wish its not me
to write it though...

**Comment by andrewbird** [comment]
I ran the synthetic benchmarks on both CPUEMU=off and CPUEMU=vm86:

Floating point (elapsed seconds)
-------------
off | vm86
----|--------
64 | 464

Integer (elapsed seconds)
-----------
off|vm86
---|-------
44|1661

So you can see the difference is significant, particularly in integer performance.

I uploaded the benchmarks here (https://gist.github.com/andrewbird/e9928742d2576b5cf958), so you can run them yourself. Of course you'll get different numbers depending upon hardware, but the difference between the modes should be evident.

**Comment by stsp** [comment]
How acceptable is this in real life?

**Comment by andrewbird** [comment]
Well, my friend runs a legacy accounting application and specifically runs 32bit Linux, just so that he can get the maximum speed with cpu=off. Other than that he updates the systems with the latest intel machine at every hardware refresh and overclocks, tweakeds DOS memory usage etc, just to get the last drop of performance. We might argue that it's one application and others might fair better, but it is a real world instance of Dosemu being used to support a legacy system which can't be changed as the source is not available. I know for a fact that he could never run with this sort of slowdown, even 10% is significant. So if the vm86() call disappears or is compile time disabled by default, he'll have stick to running an old kernel and accept that he can't have any security fixes for any exploits that appear. The main problem for him is that as he uses the latest cpu / chipsets, kernel support for any new hardware isn't going for be available.

**Comment by stsp** [comment]
OK, I run your tests, and:

C:\>FBENC~$B.EXE
Ready to begin John Walker's floating point accuracy
and performance benchmark.  1000 iterations will be made.
 
 
Measured run time in seconds should be divided by 1
to normalise for reporting results.  For archival results,
adjust iteration count so the benchmark runs about five minutes.
 
elapsed time = 0
scaled value = 0.000000
 
No errors in results.
 
C:\>ICALC~TF.EXE
Usage: prog <iterations>

**Comment by stsp** [comment]
Looks fast. :)

**Comment by stsp** [comment]
Does his program run fast enough on old 486?

**Comment by andrewbird** [comment]
Sorry, I should have said. The README shows the number of iterations to run each prog with.




Sent from Samsung tablet

-------- Original message --------
From Stas Sergeev <notifications@github.com> 
Date: 09/09/2015  12:20  (GMT+00:00) 
To stsp/dosemu2 <dosemu2@noreply.github.com> 
Cc Andrew Bird <ajb@spheresystems.co.uk> 
Subject Re: [dosemu2] What are the consequences of this kernel change? (#77) 
 
OK, I run your tests, and:

C:>FBENC~$B.EXE
Ready to begin John Walker's floating point accuracy
and performance benchmark. 1000 iterations will be made.

Measured run time in seconds should be divided by 1
to normalise for reporting results. For archival results,
adjust iteration count so the benchmark runs about five minutes.

elapsed time = 0
scaled value = 0.000000

No errors in results.

C:>ICALC~TF.EXE
Usage: prog

**Comment by stsp** [comment]
When I enable vm86 in 4.2, it oopses for me
when dosemu is starting, and the kernel hangs.
A dead horse.

**Comment by andrewbird** [comment]
Wow, then surely that's a regression kernel devs need to fix. I'm running 3.19 and vm86() is working fine there.




Sent from Samsung tablet

-------- Original message --------
From Stas Sergeev <notifications@github.com> 
Date: 09/09/2015  20:24  (GMT+00:00) 
To stsp/dosemu2 <dosemu2@noreply.github.com> 
Cc Andrew Bird <ajb@spheresystems.co.uk> 
Subject Re: [dosemu2] What are the consequences of this kernel change? (#77) 
 
When I enable vm86 in 4.2, it oopses for me
when dosemu is starting, and the kernel hangs.
A dead horse.

**Comment by andrewbird** [comment]
I doubt it was ever run on a 486, it was Pentium era. I suspect that it wasn't as fast on a Pentium as cpuemu=off on modern processors, but expectations were much lower then. One thing I do know is that it runs faster under Windows XP than even Dosemu/cpuemu=off.
It's a shame about the kernel 4.2 being a lemon, will they fix it or will it just add to their argument for removal of vm86()?

**Comment by stsp** [comment]
> One thing I do know is that it runs faster under Windows XP than even Dosemu/cpuemu=off.

Interesting.
There are 2 possible optimizations in general:
move vm86() into a separate thread in dosemu,
and make vm86() to use VME in kernel. How
the VME will speed up the purely computational
test-case is unclear, but for sure this is the only
possibility for any "ultra-light" vm86 implementation.

> It's a shame about the kernel 4.2 being a lemon, will they fix it or will
> it just add to their argument for removal of vm86()?

Worth asking if you reproduce it yourself.
My test suit for vm86() is currently very poor.
Maybe I need to install the 32bit fedora at home...

**Comment by stsp** [comment]
Could you please re-test the performance?

**Comment by andrewbird** [comment]
I ran the synthetic benchmarks with ' 2.0pre4-20150914-g0126560' on both CPUEMU=off and CPUEMU=vm86:

Floating point (elapsed seconds)
-------------
off | vm86
----|--------
64 | 544

Integer (elapsed seconds)
-----------
off|vm86
---|-------
43|1633

I guess you were looking for a speedup, but in fact floating point time is 14% longer and integer is pretty much the same at around 0.2% faster.

**Comment by stsp** [comment]
I was looking for a small speed-up for vm86(),
and that's what I see from 44 vs 43, so that's fine.
But I didn't mean a regression for cpu-emu.
Could you please see what caused this?
The interesting commits are 11ae67fd and
4011b633e, otherwise git bisect will help.

**Comment by andrewbird** [comment]
CPU 100% at idle issue
~~~
ff70979                                                                       
d57f35d                                                                        
70c19f0                                                                        
3e17290 - good for cpu idle                                                     
6c6551e - good for cpu idle                                                     
f20de5a                                                                        
86690f5 - good for cpu idle                                                     
4011b63 - bad for cpu idle - bisect suggests this is the bad commit            
73ae13c - bad for cpu idle                                                      
11ae67f                                                                        
5d7d271                                                                        
46bca6c                                                                        
f929372                                                                        
0126560 - bad CPU idle
~~~

The speed issue is less clear, as it doesn't seem to be repeatable and I can't recognise any pattern that might change the performance. Values below are elapsed seconds. The system was otherwise idle except for top and dosemu always seemed to be 100% on cpu when running the tests.
~~~
ff70979                                                                         
d57f35d                                                                         
70c19f0                                                                         
3e17290 - 490                                                                   
6c6551e - 563 484                                                               
f20de5a - 488                                                                   
86690f5 - 480                                                                   
4011b63 - 469                                                                   
73ae13c - 464                                                                   
11ae67f - 547 474                                                               
5d7d271 - 468                                                                   
46bca6c - 435                                                                   
f929372 - 547 546                                                               
0126560 - 544
~~~

**Comment by stsp** [comment]
So how can I reproduce the CPU idle problem?

**Comment by stsp** [comment]
FYI, Arjan van de Ven says that enabling by
default only means that kernel devs recommend
the wide use of some feature. Since this is not
the case for vm86, it will remain disabled even
if re-implemented.
He says distributions should enable whatever
option fits their needs, and defaults do not matter.
So he suggests us to simply ignore the change
and use the distro-provided kernels as before.

**Comment by andrewbird** [comment]
It's as simple as setting cpuemu=vm86 and starting up. It sits at the DOS prompt and uses 100% CPU.


Sent from Samsung tablet

-------- Original message --------
From Stas Sergeev <notifications@github.com> 
Date: 14/09/2015  16:52  (GMT+00:00) 
To stsp/dosemu2 <dosemu2@noreply.github.com> 
Cc Andrew Bird <ajb@spheresystems.co.uk> 
Subject Re: [dosemu2] What are the consequences of this kernel change? (#77) 
 
So how can I reproduce the CPU idle problem?

**Comment by andrewbird** [comment]
So I think the essence of the vm86() kconfig issue is buck passing by the Kernel devs. They'll set the default to disabled, because seemingly it doesn't matter. The distros will blindly follow and any user whom has a problem has two choices; recompile their own kernel or; convince each distro that their kconfig is wrong and they should enable this scary option for just some obscure app that runs obsolete programs from the 80s. Ubuntu already has it disabled for their 4.2 kernel on the as yet unreleased 15.10.
Thanks for all your help in pushing this, I do appreciate all your efforts but I think I'm coming around to the idea that getting it changed is a lost cause.
Thanks again,
Andrew

**Comment by stsp** [comment]
> It's as simple as setting cpuemu=vm86 and starting up. It sits at the DOS prompt and uses 100% CPU.

Seems like I can reproduce with FreeDOS (at home I
use PC-DOS, so haven't noticed).

> The distros will blindly follow

Only if they do not provide dosemu.

> Ubuntu already has it disabled for their 4.2 kernel on the as yet unreleased 15.10.

And stopped providing dosemu, or what?

> Thanks for all your help in pushing this, I do appreciate all your efforts but I think I'm
> coming around to the idea that getting it changed is a lost cause.

This needs to be justified.
In any case, without having vm86_light, nothing can be
changed. If vm86_light is there and dosemu2 is good enough,
we can try to convince some distros. But this is likely an
unrealistic goal given the current man-power behind dosemu2.

**Comment by andrewbird** [comment]
Wow, that seems to have fixed both the speed and the idle problems.
Here are the results
- Pre4
Idle ~4%
Test runs 493, 474, 548, 518 seconds

- Latest
Idle ~4%
Test runs 466, 470, 483, 446 seconds

It looks to me like the speed issue was lurking as far back as pre4 and only occurred randomly. Now speed seems to be more consistent and faster. Good stuff!

Regarding the manpower issue, I'd really like to see vm86_light(), but although I have worked on a few drivers in the kernel (Hso, Nozomi, Option, CDC_Ether) and had changes accepted upstream, they were mainly around USB/PCI and mostly extensions of others' work. I wouldn't know where to begin to design a new syscall that close to the processor. Given that it's already a contentious topic, a relative newbie coming in isn't going to survive for long!

We should probably close this ticket it's gone somewhat off topic now. I should have raised another for the regression, which I now regret.

**Comment by stsp** [comment]
Thanks, I'd also wonder about the native vm86()
performance improvement with this patch ($_cpuemu=off).

> I wouldn't know where to begin to design a new syscall that close to the processor.

The vm86_old syscall should be re-used. It was never
used, and sadly, because of the bug in it, I wasn't able
to make dosemu2 to use it instead of normal vm86.

> Given that it's already a contentious topic, a relative newbie coming in isn't going to survive for long!

Unless with the patch, of course. :) Oh well.

OK, lets close this, but performance results would be good.

**Comment by andrewbird** [comment]
I updated the previous comment with the cpuemu=off results. They look fine to me!

**Comment by stsp** [comment]
So no change for cpuemu=off, right?
Sigh. I thought there would be. :)

**Comment by andrewbird** [comment]
That's correct, no change for cpuemu=off.

I also ran the integer benchmarks and they are improved for cpuemu=vm86 also.

#### Pre4 - Integer (elapsed seconds)

off|vm86
---|-------
43|1633

#### Latest (4c6fd7f) - Integer (elapsed seconds)

off|vm86
---|-------
43|1516

**Comment by andrewbird** [comment]
It seems that they fixed the kconfig message, that might help in getting distros to allow vm86() http://www.spinics.net/lists/linux-tip-commits/msg30360.html

**Comment by stsp** [comment]
Yeah, this was the only positive (for our side)
result of the discussion. After that change, the
"doesn't matter at all" logic seems more plausible.
I still don't understand what you said about ubuntu 15.10 -
have they dropped dosemu?

On the positive side I've found that IRQs are
processed very slowly. I'll try to speed up things
a bit soon.

**Comment by andrewbird** [comment]
I filed a bug against kernel package on Ubuntu. https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1499089
No, dosemu 1.4.0.7 is still being packaged for Ubuntu 15.10, it's just that the kernel(4.2) on there won't have vm86() enabled.

**Comment by stsp** [comment]
Hello Andrew.
It seems your bug report got some progress,
but the reality is that the attack surface was fixed
just today, as a result of the still ongoing discussions
between me and kernel folks:
http://git.kernel.org/cgit/linux/kernel/git/tip/tip.git/commit/?id=2459ee8651dc5ab72790c2ffa99af288c7641b64
Perhaps you should make a note in the report you
opened. It is absolutely impossible to re-enable vm86()
without back-porting also this.
And please note that I've yet to test it (but you can
do that too).

**Comment by stsp** [comment]
Also it is a good time to dump other dosemu-related
problems in kernel, because Linus actively monitors
the thread and makes sure the problems are handled
properly, even for the disabled functionality. So if you
have something, please speak up. For now I am keeping
them busy with my own back-log of the problems.

**Comment by stsp** [comment]
It may be interesting if ubuntu is now affected
by the attack surface or not. The fix mentions the
offending commit. One should check if that commit
already belongs to the ubuntu kernel or not.

**Comment by andrewbird** [comment]
Ubuntu 4.2.0.19 kernel now in -proposed repos for Wily and Trusty via LTS HWE enablement stack now have support for cpuemu=off on i386(32bit x86) kernels. Stas I checked their kernel diff and that last patch seems to have been applied. https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1499089
I tested with both my floating point and integer benchmarks, and there is no slowdown evident over the 3.19 kernel I was previously using. So assuming the kernel gets moved from -proposed to the main repos we can finally put this to bed then! Thanks to everyone that added weight to the argument both here, on lkml and over at launchpad!
